src/coreclr/jit/emit.cpp

```diff
@@ -4479,6 +4610,429 @@ void emitter::emitJumpDistBind()
 #endif // DEBUG
 }
 
+#ifdef FEATURE_LOOP_ALIGN
+
+void emitter::emitLoopAlignment()
+{
+    if ((emitComp->opts.compJitAlignLoopBoundary > 16) && (!emitComp->opts.compJitAlignLoopAdaptive))
+    {
+        emitLongLoopAlign(emitComp->opts.compJitAlignLoopBoundary);
+    }
+    else
+    {
+        emitLoopAlign();
+    }
+
+    // Mark this IG as need alignment so during emitter we can check the instruction count heuristics of
+    // all IGs that follows this IG and participate in a loop.
+    emitCurIG->igFlags |= IGF_LOOP_ALIGN;
+
+    JITDUMP("Adding 'align' instruction of %d bytes in G_M%03u_IG%02u.\n", emitComp->opts.compJitAlignLoopBoundary,
+            emitComp->compMethodID, emitCurIG->igNum);
+}
+
+//-----------------------------------------------------------------------------
+//  For loopHeaderIg, find the size of the smallest possible loop that doesn't exceed maxLoopSize.
+//
+unsigned emitter::getLoopSize(insGroup* igLoopHeader, unsigned maxLoopSize)
+{
+    unsigned loopSize = 0;
+
+    for (insGroup* igInLoop = igLoopHeader; igInLoop != nullptr; igInLoop = igInLoop->igNext)
+    {
+        loopSize += igInLoop->igSize;
+        if (igInLoop->isLoopAlign())
+        {
+            // If igInLoop's next IG is a loop and needs alignment, then igInLoop should be the last IG
+            // of the current loop and should have backedge to current loop header.
```

Comment:
I thought this was simplistic instead of having portion of alignment logic in register allocator as well. With this approach, we can simply check if the last IG was also marked as IGF_LOOP_ALIGN` and if yes, then just remove the `align` instruction size from it.